### ISSUE TITLE
Return Poly.coeffs in the oder that were provided

### ISSUE THREAD

**Issue by iyanmv** [mhostetter/galois#193]
I think it would be useful to get the coefficients in the order that were provided to the constructor.

For example,

```python
import galois
from galois import GF2

# 1 + x^4 + x^6
poly = galois.Poly([1, 0, 1, 0, 0, 0, 1], field=GF2, order="asc")
poly.coeffs
```

I would expect to respect the `order="asc"` and return

```python
GF([1, 0, 1, 0, 0, 0, 1], order=2)
```
instead of

```python
GF([1, 0, 0, 0, 1, 0, 1], order=2)
```

A bit offtopic, but why is descending order the default and preferred one (for `__str__`, `Poly.string`, etc.)? Would it make sense to use same order as `numpy.polynomial.polynomial.Polynomial.coef`, which is ascending?

**Comment by mhostetter** [comment]
Hey. To be candid, the polynomial coefficient ordering was something I debated quite a bit. I settled on something and decided to wait for feedback from users.

I do think it's a good idea to allow preservation of the coefficient ordering. Perhaps that could be accomplished by adding a `Poly.order` property that holds the `order` kwarg value used at construction, either `"asc"` or `"desc"`. Then, `Poly.coeffs` would return the coefficients in `Poly.order` order.

Are you also suggesting that if the new property `Poly.order = "asc"` that `str()` and `Poly.string` should return the string representation of the polynomial with ascending degrees? For instance, `Poly(1 + x^4 + x^6, GF(2))` and `1 + x^4 + x^6` when `Poly.order = "asc"`?

As to the default, again I debated that. I'm open to changing the default to ascending. I've seen that in other libraries ascending is common.

To be clear up a typo (I think) from your example... Since you specified `order="asc"`, that polynomial is `1 + x^2 + x^6` not `1 + x^4 + x^6`. Do you agree?

```python
In [1]: import galois

In [2]: poly = galois.Poly([1, 0, 1, 0, 0, 0, 1], order="asc")

In [3]: poly
Out[3]: Poly(x^6 + x^2 + 1, GF(2))

# Currently, coefficients are always in descending order
In [4]: poly.coeffs
Out[4]: GF([1, 0, 0, 0, 1, 0, 1], order=2)
```

**Comment by iyanmv** [comment]
Thanks for answering so quickly!

> I do think it's a good idea to allow preservation of the coefficient ordering. Perhaps that could be accomplished by adding a Poly.order property that holds the order kwarg value used at construction, either "asc" or "desc". Then, Poly.coeffs would return the coefficients in Poly.order order.

Okay, let me have a closer look at the code and maybe I can implement this based on your notes and do a pull request.

> Are you also suggesting that if the new property Poly.order = "asc" that str() and Poly.string should return the string representation of the polynomial with ascending degrees? For instance, Poly(1 + x^4 + x^6, GF(2)) and 1 + x^4 + x^6 when Poly.order = "asc"?

Yes, that's what I meant.

> As to the default, again I debated that. I'm open to changing the default to ascending. I've seen that in other libraries ascending is common.

I think it would make sense to use the same order as NumPy by default, but allow the user to change it. I also realized that each library is using a different order, and it can get confusing when you work with all of them. SymPy, for example, also uses descending order by default. I asked because I wanted to understand if there was any advantage of of using descending over ascending, but I guess it's just something you decided, and then you have to stick with it and be consistent.

> To be clear up a typo (I think) from your example... Since you specified order="asc", that polynomial is 1 + x^2 + x^6 not 1 + x^4 + x^6. Do you agree?

Yes, yes, I agree! That was a typo :facepalm:

**Comment by iyanmv** [comment]
By the way, is your goal in the long term to add this to NumPy or to keep it as a separate module? I find this module very useful, and I discovered it today by mistake while looking for a different BCH code implementation. Don't you think it would fit inside NumPy?

**Comment by mhostetter** [comment]
Thanks for the offer to help, but I'd wait a bit. The `Poly` class is a bit of a mess right now and I've been meaning to refactor it. Also, a lot of the underlying private helper functions presume the descending order of coefficients. So just adding the `order` property and returning `coeffs` based on that won't be sufficient.

Additionally, as I'm thinking in the 5 minutes since writing my last post, I'm concerned about creating "statefulness" inside a `Poly` object.

For instance, take two polynomials `a` and `b`. `a` was created with `order="asc"` and `b` was created with `order="desc"`. Calling `a.coeffs` and `b.coeffs` necessitates knowledge of the creation order (or current value of `a.order` and `b.order`). It may cause users to need to wrap `a.coeffs` calls in `if a.order == "asc"`, or something similar.

Furthermore, given `c = a + b`, what should `c.order` be? The only sensible answer I can think of would be the default order (whatever that ends up being). But a user like you who creates polynomial `a` in ascending order and adds it to polynomial `b` that was created elsewhere in descending order and gets resulting polynomial `c` in the *default* descending order, I think that user would be confused.

I'd like to avoid statefulness, wherever possible.

> By the way, is your goal in the long term to add this to NumPy or to keep it as a separate module? I find this module very useful, and I discovered it today by mistake while looking for a different BCH code implementation. Don't you think it would fit inside NumPy?

I'm open to integrating this code somehow into base NumPy if the developers thought it was appropriate there. I doubt they would, however. My tact was to create a `ndarray` subclass that can be easily "switched in" or "switcched out" using `.view()` that operates over a different field than the reals, specifically finite fields. However, there are many other fields that *could* be supported. [`quaternionic`](https://github.com/moble/quaternionic) did something similar for quaternion arrays. I imagine if NumPy wanted to integrate something like this, they would want to allow for more fields than just finite fields. But again, I'm open to it if the devs think it's prudent.

**Comment by iyanmv** [comment]
Okay, sure, sure, I can wait!

What if `Poly.coeffs` instead of a property is a method with an order parameter with the same default value as the constructor (whatever it is at the end)?

```python
import galois
from galois import GF2

a = galois.Poly([1, 1, 0, 1], field=GF2, order="asc")  # a(x) = 1 + x + x^3
b = galois.Poly([0, 0, 1, 0], field=GF2)  # b(x) = x
c = a + b  # 1 + x^3

a.coeffs()
GF([1, 0, 1, 1], order=2)

a.coeffs(order="asc")
GF([1, 1, 0, 1], order=2)

b.coeffs()
GF([1, 0], order=2)

b.coeffs(order="asc")
GF([0, 1], order=2)

c.coeffs
GF([1, 0, 0, 1], order=2)

c.coeffs(order="asc")
GF([1, 0, 0, 1], order=2)
```

So internally, all the operations are done, let's say assuming a descending order, but the user is free to create `Poly` objects in both orders as well as get back `ndarrays` in the desired order.

Also, maybe is better to change `order` to `arrangement` or `ordering`, for example, to avoid the confusion with `Field.order`.

**Comment by mhostetter** [comment]
> What if Poly.coeffs instead of a property is a method with an order parameter with the same default value as the constructor (whatever it is at the end)?

I thought of something similar in the beginning. And, actually, I once had `Poly.coeffs_desc` and `Poly.coeffs_asc` to allow easy access to either, which is similar to your `Poly.coeffs("desc")` and `Poly.coeffs("asc")` suggestion.

I am hesitant to have one attribute accessible via method accessor and the rest via conventional property decorator. For instance, `Poly.coeffs` and `Poly.degrees` are one-for-one pairs. `Poly.coeffs()` and `Poly.degrees` is less intuitive. And when `Poly.coeffs("asc"),` then `Poly.degrees` is no longer in corresponding order with the coefficients. Perhaps, if one were a method accessor, they both should be. But I would also, then, like to apply that to `Poly.nonzero_coeffs()` and `Poly.nonzero_degrees()`.

Your suggestion would have users likely calling `Poly.coeffs()` (for the default) or `Poly.coeffs("asc")` (for the non-default). That's essentially the same as the current API: `Poly.coeffs` (the default) or `Poly.coeffs[::-1]` (for the non-default).

I'm not quite sure what to do yet. I'm still thinking about it and open to more thoughts and suggestions.

> Also, maybe is better to change order to arrangement or ordering, for example, to avoid the confusion with Field.order.

I had this same thought when I added that kwarg, however I couldn't come up with a better, succinct variable name. `ordering` might be better. I was hoping with the `field` kwarg right there, it would be clear that the field is defined there and not with `order`. But point taken, I can definitely see where confusion could arise.

**Comment by iyanmv** [comment]
> I'm not quite sure what to do yet. I'm still thinking about it and open to more thoughts and suggestions.

The more I think about it, the more confused I am about what I would like to have. I think for the moment, I am happy with `Poly.coeffs[::-1]`. If I think of a better way, I will let you know.

**Comment by mhostetter** [comment]
@iyanmv I was thinking about this some more... What about the below proposal to solve the coefficient ordering issue raised here *and* the coefficient size issue raised in #197.

## Proposal
Keep all current properties (namely `Poly.coeffs`) as is. Add an additional method accessor `Poly.coefficients(size=None, order="desc")`. `size=None` defaults to `degree + 1`. So by default, `poly.coefficients() == poly.coeffs`.

This would allow a user like you to access the coefficients in ascending order with `poly.coefficients(order="asc")` in a more "readible" way than `poly.coeffs[::-1]`. It would also allow a user to return the coefficients in a fixed-size array, e.g. `poly.coefficients(32)` from this [hypothetical](https://github.com/mhostetter/galois/issues/197#issuecomment-962117683).

It would be nice if I could overload the `Poly.coeffs` property and the `Poly.coeffs()` method accessor, but I don't believe that's allowable in Python, hence why I chose the `Poly.coefficients()` name.

**Comment by iyanmv** [comment]
> Keep all current properties (namely `Poly.coeffs`) as is. Add an additional method accessor `Poly.coefficients(size=None, order="desc")`. `size=None` defaults to `degree + 1`. So by default, `poly.coefficients() == poly.coeffs`.

This sounds really good!

> This would allow a user like you to access the coefficients in ascending order with `poly.coefficients(order="asc")` in a more "readible" way than `poly.coeffs[::-1]`. It would also allow a user to return the coefficients in a fixed-size array, e.g. `poly.coefficients(32)` from this [hypothetical](https://github.com/mhostetter/galois/issues/197#issuecomment-962117683).
> 
> It would be nice if I could overload the `Poly.coeffs` property and the `Poly.coeffs()` method accessor, but I don't believe that's allowable in Python, hence why I chose the `Poly.coefficients()` name.

Yeah, I also don't think is possible. But I guess it's not a big issue to have both `coeffs` and `coefficients()`, don't you think?

**Comment by mhostetter** [comment]
Yeah, I agree. I'll incorporate this idea into my next PR.

**Comment by mhostetter** [comment]
@iyanmv do you think this issue is resolved given #210?

**Comment by iyanmv** [comment]
Yes, I think so. Thanks!
